### PR TITLE
Ensure the radial setting of PSF models are written out by the save_all command 

### LINKED ISSUE #2470: save_all does not save the radial flag for 1D PSFs
Found when developing #2447 

If you do something like

```python
load_psf("p1", "foo.dat")
set_psf(p1)
p1.radial = 1
```

then the `save_all` output is just

```
######### Set Model Components and Parameters

load_psf("p1", "foo.dat")


######### Associate PSF models with the datasets

set_psf(1, p1)


```

That is, it does not record the change to `radial` (in a real-world case there would be some changes to `p1` but not `radial`).

### LINKED ISSUE #2506: save_all: does not handle load_conv
This is with CIAO 4.18 but the code has been like this for a while;

```
sherpa
-----------------------------------------------------
Welcome to Sherpa: CXC's Modeling and Fitting Package
-----------------------------------------------------
Sherpa 4.18.0+2.ga0c5211a

Python 3.12.12 | packaged by conda-forge | (main, Oct 13 2025, 14:47:45) [Clang 19.1.7 ]
Type 'copyright', 'credits' or 'license' for more information
IPython 9.13.0 -- An enhanced Interactive Python. Type '?' for help.

IPython profile: sherpa
Tip: Use `ipython --help-all | less` to view all the IPython configuration options.
Using matplotlib backend: macosx

sherpa In [1]: load_conv("bob", gauss1d.g1)

sherpa In [2]: from io import StringIO

sherpa In [3]: buff = StringIO(); save_all(buff)

sherpa In [4]: clean()

sherpa In [5]: exec(buff.getvalue())
ArgumentErr: 'convolutionkernel' is not a valid model type

sherpa In [6]:
```

This is because the serialization stores something like

```
create_model_component("convolutionkernel", "bob")
load_conv("bob", "gauss1d.g1")
```

where the `create_model_component` line is wrong and can just be dropped.

### PR DESCRIPTION
**Pull request by DougBurke**
# Summary

Ensure that the radial setting of a PSF model and error-estimation options are saved in the save_all output. The method options are now only written out when explicitly set. Fix #2470. Fix #2506.

# Details

This came out of #2447. We have 

- add a test or two to cover some corner cases, including #2470, #2504, #2506
- some minor code clean up
- fix some of the bugs, including #2470 and #2506 (but not #2504)

Unfortunately - for this code at least - the PSF model class 

- doesn't follow other classes and have a `.pars` field (there are reasons for this, it's just why this issue happened)
- has some "parameters" only appear after associated with a dataset

which means we always risk the "missing a field" case (even after this PR; I have thought about trying to make it "generic" so the special-case code would not be needed, but it does not feel like it's worth it at this time given the extra complexity likely required).

A change is made to support properly restoring `load_conv` calls - following how we do it for `load_psf`. This is issue #2506.

Doing a little bit of type checking with `pyright` made me some changes that are either just to make it easier for the type checkers (but probably also make the code a bit cleaner as there is less "boolean blindness") or actually avoid an issue (where the psf or convolution kernel is not set) which is extremely difficult to pull off for users of the UI layer and so no test is added.

# Example

Before this PR, not that the `save_all` output does not mention the `radial` parameter:

```
>>> from sherpa.astro import ui
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find ds9 on your PATH'
>>> from io import StringIO
>>> ui.load_psf("bob", ui.gauss1d.g1)
>>> print(bob)
psfmodel.bob
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   bob.kernel   frozen   gauss1d.g1
   bob.radial   frozen            0            0            1           
   bob.norm     frozen            1            0            1           
>>> bob.radial = 1
>>> print(bob)
psfmodel.bob
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   bob.kernel   frozen   gauss1d.g1
   bob.radial   frozen            1            0            1           
   bob.norm     frozen            1            0            1           
>>> buff = StringIO()
>>> ui.save_all(buff)
>>> print(buff.getvalue())
import numpy
from sherpa.astro.ui import *


######### Set Statistic

set_stat("chi2gehrels")


######### Set Fitting Method

set_method("levmar")

set_method_opt("epsfcn", 1.1920928955078125e-07)
set_method_opt("factor", 100.0)
set_method_opt("ftol", 1.1920928955078125e-07)
set_method_opt("gtol", 1.1920928955078125e-07)
set_method_opt("maxfev", None)
set_method_opt("numcores", 1)
set_method_opt("verbose", 0)
set_method_opt("xtol", 1.1920928955078125e-07)


######### Set Model Components and Parameters

create_model_component("gauss1d", "g1")
g1.integrate = True

g1.fwhm.default_val = 10.0
g1.fwhm.default_min = 1.1754943508222875e-38
g1.fwhm.default_max = 3.4028234663852886e+38
g1.fwhm.val     = 10.0
g1.fwhm.min     = 1.1754943508222875e-38
g1.fwhm.max     = 3.4028234663852886e+38
g1.fwhm.units   = ""
g1.fwhm.frozen  = True

g1.pos.default_val = 0.0
g1.pos.default_min = -3.4028234663852886e+38
g1.pos.default_max = 3.4028234663852886e+38
g1.pos.val     = 0.0
g1.pos.min     = -3.4028234663852886e+38
g1.pos.max     = 3.4028234663852886e+38
g1.pos.units   = ""
g1.pos.frozen  = True

g1.ampl.default_val = 1.0
g1.ampl.default_min = -3.4028234663852886e+38
g1.ampl.default_max = 3.4028234663852886e+38
g1.ampl.val     = 1.0
g1.ampl.min     = -3.4028234663852886e+38
g1.ampl.max     = 3.4028234663852886e+38
g1.ampl.units   = ""
g1.ampl.frozen  = True

load_psf("bob", "gauss1d.g1")


>>> 
```

The new output is

```
>>> >>> print(buff.getvalue())
import numpy
from sherpa.astro.ui import *


######### Set Statistic

set_stat("chi2gehrels")


######### Set Fitting Method

set_method("levmar")

set_method_opt("epsfcn", 1.1920928955078125e-07)
set_method_opt("factor", 100.0)
set_method_opt("ftol", 1.1920928955078125e-07)
set_method_opt("gtol", 1.1920928955078125e-07)
set_method_opt("maxfev", None)
set_method_opt("numcores", 1)
set_method_opt("verbose", 0)
set_method_opt("xtol", 1.1920928955078125e-07)


######### Set Model Components and Parameters

create_model_component("gauss1d", "g1")
g1.integrate = True

g1.fwhm.default_val = 10.0
g1.fwhm.default_min = 1.1754943508222875e-38
g1.fwhm.default_max = 3.4028234663852886e+38
g1.fwhm.val     = 10.0
g1.fwhm.min     = 1.1754943508222875e-38
g1.fwhm.max     = 3.4028234663852886e+38
g1.fwhm.units   = ""
g1.fwhm.frozen  = True

g1.pos.default_val = 0.0
g1.pos.default_min = -3.4028234663852886e+38
g1.pos.default_max = 3.4028234663852886e+38
g1.pos.val     = 0.0
g1.pos.min     = -3.4028234663852886e+38
g1.pos.max     = 3.4028234663852886e+38
g1.pos.units   = ""
g1.pos.frozen  = True

g1.ampl.default_val = 1.0
g1.ampl.default_min = -3.4028234663852886e+38
g1.ampl.default_max = 3.4028234663852886e+38
g1.ampl.val     = 1.0
g1.ampl.min     = -3.4028234663852886e+38
g1.ampl.max     = 3.4028234663852886e+38
g1.ampl.units   = ""
g1.ampl.frozen  = True

load_psf("bob", "gauss1d.g1")
bob.radial = 1.0
bob.norm = 1.0


>>>
```

So, after the `load_psf` call we now have `radial` and `norm` settings.

For the `load_conv` case we have:

```
>>> from sherpa.astro import ui
WARNING: imaging routines will not be available, 
failed to import sherpa.image.ds9_backend due to 
'RuntimeErr: DS9Win unusable: Could not find xpaget on your PATH'
WARNING: failed to import sherpa.astro.xspec; XSPEC models will not be available
>>> from io import StringIO
>>> ui.load_conv("c1", ui.gauss1d.g1)
>>> buff = StringIO(); ui.save_all(buff)
>>> ui.clean()
>>> print(c1)
Traceback (most recent call last):
  File "<python-input-5>", line 1, in <module>
    print(c1)
          ^^
NameError: name 'c1' is not defined
```

## before

```
>>> exec(buff.getvalue())
Traceback (most recent call last):
  File "<python-input-6>", line 1, in <module>
    exec(buff.getvalue())
    ~~~~^^^^^^^^^^^^^^^^^
  File "<string>", line 56, in <module>
  File "<string>", line 1, in create_model_component
  File "/home/dburke/sherpa/sherpa-3/sherpa/ui/utils.py", line 6686, in create_model_component
    raise ArgumentErr('badtype', typename)
sherpa.utils.err.ArgumentErr: 'convolutionkernel' is not a valid model type
```

## after

```
>>> exec(buff.getvalue())
>>> print(c1)
Convolution Kernel:
gauss1d.g1
   Param        Type          Value          Min          Max      Units
   -----        ----          -----          ---          ---      -----
   g1.fwhm      thawed           10  1.17549e-38  3.40282e+38           
   g1.pos       thawed            0 -3.40282e+38  3.40282e+38           
   g1.ampl      thawed            1 -3.40282e+38  3.40282e+38           
```

### CODE REVIEW

**Comment by DougBurke** [comment]
Why does conda installing

```
  added / updated specs:
    - astropy
    - gcc_linux-64=14.2
    - gfortran_linux-64
    - gxx_linux-64=14.2
    - numpy=2.3.2
    - pip
    - python=3.13
```

lead to matplotlib also being installed? Looks like it's an astropy dependency, which messes up our CI builds a little bit ....

**Comment by hamogu** [comment]
Matplotlib is an optional astropy dependency (https://github.com/astropy/astropy/blob/main/pyproject.toml#L45), not a required one, but maybe the recipe in conda includes all astropy dependencies that are in the "recommended" group equivalent to `pip install "astropy[recommended]"` (which, as the name implies, is what astropy recommends for all users, so it would make sense to package it in conda like that). We could switch astropy to a pip install instead of conda, then we're guaranteed to get only the strictly required dependencies and since astropy builds wheels it won't take any longer than a conda install.

**Comment by DougBurke** [comment]
Rebased now that #2471 has been fixed. The RTD docs will fail because of #2479

**Comment by DougBurke** [comment]
Rebased as #2473 was merged to fix #2479

**Comment by DougBurke** [comment]
Rebased to remove some clean-up code that is not directly related to the bug fix and may slightly conflict with some other PRs

**Comment by DougBurke** [comment]
Rebased to pick up #2484 and to fix up a few stupid typos in the commit messages. That's not to say there aren't still some hiding away.

File: sherpa/astro/ui/serialize.py [hunk 1]
```diff
@@ -1001,6 +998,125 @@ def _handle_usermodel(out: OutType,
...
+    # If we have a name "aa.bb" then return "bb" otherwise
+    # return the name unchanged.
+    #
+    name = mod.name
+    try:
+        return name.split(".")[1]
+    except IndexError:
+        return name
+
+
```
Comment (hamogu) [inline sherpa/astro/ui/serialize.py:1013]: Wouldn't it be easier to make this a one-liner and inline it where needed?
```
variable_I_want_to_set = mod.names.split(".")[-1]
```
I always think that the mental overhead for a reader to go and look at what exactly a private function does is higher than reading the one line of standard Python.

Comment (hamogu) [reply sherpa/astro/ui/serialize.py:1013]: I also note that heuristics like this can be dangerous, see #1384. Here, the failure mode is the opposite: What if my model is called "xsapec.src1.cold"?
I don't see a way around corner cases like that when you need to rely on heuristics, and we can probably live with it, but it's certainly not ideal.

File: sherpa/astro/ui/serialize.py [hunk 2]
```diff
@@ -1001,6 +998,125 @@ def _handle_usermodel(out: OutType,
...
+        _output(out, par_attributes)
+        if par_linkstr is not None:
+            links.append(par_linkstr)
+
+    # If the model is a PSFModel then there are a number of
+    # attributes we want to set, but they are not stored in the
+    # .pars attribute. Not all are Parameter objects.
+    #
+    # Drop the "kernel" field as it should not be user settable.  Do
+    # not worry about other parameter attrbutes such as a frozen flag
```
Comment (hamogu) [inline sherpa/astro/ui/serialize.py:1092]:
```suggestion
    # not worry about other parameter attributes such as a frozen flag
```

File: sherpa/astro/ui/serialize.py [hunk 3]
```diff
@@ -1144,123 +1195,136 @@ def _save_psf_components(out: OutType, state: SessionType) -> None:
...
 
-    ids = state.list_data_ids()
-    if len(ids) == 0:
-        return
+    # If a data set has a source model associated with it,
+    # set that here -- try to distinguish cases where
+    # source model is different from whole model.
+    # If not, do nothing.
+    #
+    with suppress(Exception):
```
Comment (hamogu) [inline sherpa/astro/ui/serialize.py:1211]: I don't quite understand the structure of this. The try..except blocks in the next few lines already catch all exceptions without qualification. So why is this nested inside an additional "suppress"?

There are few lines that are not in individual try..except blocks, such as `have_source = the_source is not None`, but I don't see how these could possibly fail in a way we simply want to ignore.

Then there are the `_output` and `_outputnl` lines at the end of the block, but in those cases, I assume that we DO want to know if it fails, right?
But I'm probably missing something here since the `suppress` is a new addition in this PR, so there much have been some intention behind it that I just don't get.

File: sherpa/astro/ui/serialize.py [hunk 4]
```diff
@@ -1144,123 +1195,136 @@ def _save_psf_components(out: OutType, state: SessionType) -> None:
...
+                     id: IdType
+                     ) -> None:
+    """Create the save_bkg_source/model/full_model line."""
+
+    # Set background models (if any) associated with backgrounds
+    # tied to this data set -- if none, then do nothing.  Again, try
+    # to distinguish cases where background "source" model is
+    # different from whole background model.
+    #
+    with suppress(Exception):
```
Comment (hamogu) [inline sherpa/astro/ui/serialize.py:1262]: Same here. What is it that we actually want to suppress?

**Review by hamogu — CHANGES_REQUESTED** [review]

File: sherpa/astro/ui/serialize.py (hunk 1 above)
Comment (DougBurke) [reply sherpa/astro/ui/serialize.py]: There's also the side of "making sure we use the same approach" throughout the code base. Now, in this case we only call it in two places, but they are separated. 

I primarily wrote this because of the exception handling, but your approach (using -1 to grab the last element rather than my code which uses 1, hence the need for the exception check) does make it easier to have as a one liner.

File: sherpa/astro/ui/serialize.py (hunk 3 above)
Comment (DougBurke) [reply sherpa/astro/ui/serialize.py]: So, for both this and the next point, this PR re-organized code, so that what was once a big routine got split into several ones, and then they both changed

```python
    try:
        do_some_stuff_tm()
    except Exception:
        return # or pass
```

into

```python
    with suppress(Exception):
        do_some_stuff_tm()
```

I didn't look closely at the code in `do_some_stuff....` to see whether the safety check was needed. I've now fixed up both to avoid the check (although in the second case there is actually one place where an error could get thrown, so that is now explicitly guarded).

**Review by hamogu — APPROVED** [review]
